lib/diskoverflow/common.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Package diskoverflow provides several data container types which are limited
+// in their memory usage. Once the total memory limit is reached, all new data
+// is written to disk.
+// Do not use any instances of these types concurrently!
+package diskoverflow
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+const concurrencyMsg = "iteration in progress - don't modify or start a new iteration concurrently"
+
+const OrigDefaultOverflowBytes = 16 << protocol.MiB
+
+var (
+	defaultOverflow    = OrigDefaultOverflowBytes
+	defaultOverflowMut = sync.Mutex{}
+)
+
+// Sets the default limit of when data is written to disks. Can be overruled
+// on individual instances via their SetOverflowBytes method.
+// A value of bytes =< 0 means no disk spilling will ever happen.
+func SetDefaultOverflowBytes(bytes int) {
+	defaultOverflowMut.Lock()
+	defaultOverflow = bytes
+	defaultOverflowMut.Unlock()
+}
+
+func DefaultOverflowBytes() int {
+	defaultOverflowMut.Lock()
+	defer defaultOverflowMut.Unlock()
+	return defaultOverflow
+}
+
+// Value must be implemented by every type that is to be stored in a disk spilling container.
+// Un-/Marshal returning an error will result in a panic.
+// This uses methods from proto interfaces such that existing types can be reused.
+type Value interface {
+	Marshal() ([]byte, error)
+	Unmarshal([]byte) error
+	Reset() // To make an already populated Value ready for Unmarshal
+	ProtoSize() int
+}
+
+// copyValue copies the content from src to dst. Src and dst must be pointers
+// to the same underlying types, otherwise this will panic.
+func copyValue(dst, src Value) {
+	dstv := reflect.ValueOf(dst).Elem()
+	srcv := reflect.ValueOf(src).Elem()
+	dstv.Set(srcv)
```

Comment:
Should this do a reset on the destination? Or is that not needed? If it's not needed, why are we calling it in quite a few places?

Reply:
Depends on the destination: Some types will not have any problems with unmarshalling over existing data and if there are problems, the destination might already be "pristine". That's why I left it up to the caller to make sure it works.